# Canadian Space Agency Challenge -- RADI-N2

This is the file used to clean and organize the datasets from the Canadian Space Agency Challence regarding radiation.
The file `Radi-N2 -ISS data.csv` contains the information provided by the Canadian austronauts on the International Space Station(ISS). The data was collected for each 7th day mentioned in the interval.

The austronauts measured the neutron values on a 7-day period. They used a neutron dosimeter.

In [1]:
# import panda library
import pandas as pd

In [2]:
# read the file
df = pd.read_csv("Radi-N2 -ISS data.csv")

In [3]:
# Let's study how the data looks like
df.head(7)

ROBERT THIRSK MISSION DATA            Unnamed: 1  \
0                             Start-Date(YYYY/MM/DD)  End-Date(YYYY/MM/DD)   
1                                         2009-09-06            2009-09-13   
2                                         2009-10-03            2009-10-10   
3                                         2009-11-09            2019-11-16   
4                                                NaN                   NaN   
5  *Sensitivity calculation : Includes space cali...                   NaN   
6  It should be noted that the results from space...                   NaN   

                      Unnamed: 2       Unnamed: 3            Unnamed: 4  \
0  Measurement-Duration(minutes)  Detector-Number  Sensitivity(b/mrem)*   
1                           9872       A92 (1415)                  0.86   
2                          10793       A92 (1415)                  0.86   
3                          10069       A92 (1415)                  0.86   
4                            NaN              NaN                   NaN   
5                            NaN              NaN                   NaN   
6                            NaN              NaN                   NaN   

     Unnamed: 5                 Unnamed: 6  
0  Bubble Count  Calculated-Exposure(mrem)  
1           107                        124  
2           141                        164  
3            65                         76  
4           NaN                        NaN  
5           NaN                        NaN  
6           NaN                        NaN

The next file is provided by students across Canada who measured the neutron concentration in their schools.

In [4]:
studs = pd.read_csv("Radi-N2 -Schools project data.csv")
studs.head(7)

Identifier                Collection Date  \
0  8163f602-9db5-4c01-9045-d7d997315c82  2018-12-18T12:21:00.000-05:00   
1  8163f602-9db5-4c01-9045-d7d997315c82  2019-01-16T12:46:00.000-05:00   
2  1f129390-03f2-429b-abc6-66cfee9bf519  2019-01-16T14:16:00.000-04:00   
3  2e8106c0-d481-446b-bbbb-c32740eef411  2019-02-19T07:30:00.000-05:00   
4  df5eceb7-3194-4e9a-86e0-b5e9f1ac8413  2019-02-19T07:30:00.000-05:00   
5  6648c712-0630-4539-af6d-f59cd463ae54  2019-02-19T07:30:00.000-05:00   
6  5f93d214-802f-415b-8551-862e28c19789  2019-01-16T14:00:00.000-08:00   

   Number of bubbles  Bubble Detector ID  Sensitivity (b/mrem)  \
0                  3            18051217                    44   
1                  0            18129251                    41   
2                  1            18047344                    47   
3                  0            18051554                    47   
4                  0            18051554                    47   
5                  0            18051554                    47   
6                  1            18046220                    53   

   Calculated Exposure (mrem)                   City Province/Territory  
0                    0.068000                 London            Ontario  
1                    0.000000                 London            Ontario  
2                    0.021277  Florenceville-Bristol      New Brunswick  
3                    0.000000              Longueuil             Quebec  
4                    0.000000              Longueuil             Quebec  
5                    0.000000              Longueuil             Quebec  
6                    0.018868             Salmon Arm   British Columbia

In [5]:
studs = studs[(studs.T != 0).all()]

In [6]:
studs.shape

(557, 8)

### How many identifiers per cities?

In [7]:
# group by City
count_ids = studs.groupby(['City']).count()
count_ids['Identifier']

City
Amqui                                      9
Belleville                                 8
Brampton                                   7
Bridgewater                               15
Calgary                                   18
Camrose                                    2
Channel-Port Aux Basques                   6
Chilliwack                                 5
Coquitlam                                  7
Dartmouth                                  1
Dartmouth N.-E.                            5
Deep River                                 5
Dryden                                     4
East Angus                                 9
Eganville                                  3
Embrun                                     1
Florenceville-Bristol                      5
Gloucester                                 1
Grand Falls-Windsor                        4
Grimsby                                    5
Guelph                                     2
Hamilton                                   3
Hanov

### What is the minimum and maximum bubbles count? What about calculated exposure?

In [8]:
studs.min()

Identifier                    009d1d48-27ab-4446-b20a-b67855ff8e8b
Collection Date                      2018-12-18T12:21:00.000-05:00
Number of bubbles                                                1
Bubble Detector ID                                         1805135
Sensitivity (b/mrem)                                            41
Calculated Exposure (mrem)                                  0.0149
City                                                         Amqui
Province/Territory                                         Alberta
dtype: object

In [9]:
studs.max()

Identifier                    fb1afc76-5b15-4ded-9e8f-d38213229c26
Collection Date                      2019-06-08T12:30:00.000-04:00
Number of bubbles                                               28
Bubble Detector ID                                        19051231
Sensitivity (b/mrem)                                            68
Calculated Exposure (mrem)                                0.666667
City                                                      Winnipeg
Province/Territory                                    Saskatchewan
dtype: object

- The minimum bubbles count is 1 and maximum is 28.
- The minimum exposure is 0.0149 mrem and maximum exposure is 0.666667 mrem.

### How many have identified 5 or less bubbles?

In [10]:
import numpy as np
# use np.where to identifier which indices have number of bubbles less than 5
np.where(studs['Number of bubbles']<5)

(array([  0,   1,   2,   3,   4,   5,   8,   9,  10,  11,  12,  13,  14,
         15,  16,  18,  19,  20,  22,  23,  24,  25,  26,  27,  28,  34,
         35,  36,  37,  39,  41,  43,  45,  47,  48,  49,  50,  51,  53,
         55,  56,  57,  58,  59,  60,  62,  63,  64,  65,  66,  67,  68,
         69,  70,  71,  72,  73,  74,  77,  78,  79,  80,  81,  82,  83,
         85,  86,  87,  88,  89,  91,  92,  93,  94,  96,  97,  98,  99,
        100, 101, 102, 103, 104, 105, 107, 108, 109, 110, 111, 112, 114,
        115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127,
        128, 129, 130, 131, 132, 133, 135, 141, 142, 145, 146, 147, 148,
        149, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163,
        164, 165, 166, 167, 168, 169, 170, 172, 177, 178, 179, 180, 181,
        182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194,
        195, 196, 197, 199, 202, 204, 205, 206, 207, 208, 209, 210, 211,
        212, 213, 214, 215, 216, 217, 218, 219, 220

There are about 444 schools that reported less than 5 bubbles. This is quite a huge drop in numbers.

## How many unique identifiers?

In [11]:
# How many identifiers 
iden = studs['Identifier'].unique()
len(iden)

105

Since there are only 105 unique identifiers, let's take a look at a random city and how were the collection dates for it. 

_Recall_: they were supposed to measure at 7-day intervals.

In [12]:
print(studs.loc[studs['City'] == 'Petawawa'])


                               Identifier                Collection Date  \
126  959484ae-c895-403a-83df-59404e44b9f6  2019-02-21T07:52:00.000-05:00   
157  959484ae-c895-403a-83df-59404e44b9f6  2019-02-22T08:25:00.000-05:00   
402  959484ae-c895-403a-83df-59404e44b9f6  2019-03-19T09:00:00.000-04:00   
403  959484ae-c895-403a-83df-59404e44b9f6  2019-03-20T09:30:00.000-04:00   
404  959484ae-c895-403a-83df-59404e44b9f6  2019-03-21T09:16:00.000-04:00   
415  959484ae-c895-403a-83df-59404e44b9f6  2019-03-22T10:04:00.000-04:00   

     Number of bubbles  Bubble Detector ID  Sensitivity (b/mrem)  \
126                  1            17170336                    56   
157                  1            17170336                    56   
402                  1            17170336                    56   
403                  2            17170336                    56   
404                  2            17170336                    56   
415                  1            17170336                 

In [13]:
# find the cities that have a 7 day range. Plot some of them. with unique identifiers
studs.columns

Index(['Identifier', 'Collection Date', 'Number of bubbles',
       'Bubble Detector ID', 'Sensitivity (b/mrem)',
       'Calculated Exposure (mrem)', 'City', 'Province/Territory'],
      dtype='object')

### What locations identified more than 5 bubbles?

In [14]:
# see the collection dates
# studs['Collection Date']

grouped_df = studs.groupby(['City','Identifier']).mean()

np.where(grouped_df['Number of bubbles']>5)

(array([11, 17, 25, 41, 45, 50, 54, 58, 73, 83, 85, 91, 94, 98],
       dtype=int64),)

We got 14 locations where they got more than 5 bubbles. So average for them will be much higher than the rest of Canada. 
Are those locations close to one another? Or are they just human error? 

In [15]:
aprox_values = round(grouped_df['Number of bubbles'])
aprox_values

City                                    Identifier                          
Amqui                                   67837e41-158f-4f50-b569-b342e84d1233     2.0
Belleville                              89d487a5-bd45-45b6-8118-13953c307150     1.0
Brampton                                6c107aaf-de6a-404a-9864-a8a708a852e4     2.0
                                        af8ce2f7-8567-4534-bbf7-7b2892b535e0     2.0
Bridgewater                             02af2a79-c8f7-4353-9163-8755a007df08     3.0
                                        0b35e95b-f234-4bdd-b0b0-4048847dfe28     2.0
Calgary                                 7cda7462-6d8d-439c-b482-5bd6d768352a     3.0
                                        807d5dcd-3485-4f3c-87e4-ae51dd391dd6     3.0
                                        a80a5290-4bab-4317-a175-4d0a017ac54c     3.0
Camrose                                 009d1d48-27ab-4446-b20a-b67855ff8e8b     1.0
                                        f454b9ad-9cbb-4817-a4c5-b0eddced5

##  GOAL: Find identifiers that do not have 7 days and normalize

# 1. Organize by identifier and collection date
Organize them by identifier and collection date.

In [16]:
# sort the days for each identifier
studs_sorted_df = studs.sort_values(['Identifier','Collection Date'], ascending = [True,True])
studs_sorted_df.head(5)

Identifier                Collection Date  \
58  009d1d48-27ab-4446-b20a-b67855ff8e8b  2019-01-30T08:52:00.000-07:00   
45  02127c7f-69ce-41ee-ae58-402c405b5e49  2019-01-22T11:00:00.000-05:00   
46  02127c7f-69ce-41ee-ae58-402c405b5e49  2019-01-23T11:00:00.000-05:00   
47  02127c7f-69ce-41ee-ae58-402c405b5e49  2019-01-23T11:00:00.000-05:00   
48  02127c7f-69ce-41ee-ae58-402c405b5e49  2019-01-24T11:00:00.000-05:00   

    Number of bubbles  Bubble Detector ID  Sensitivity (b/mrem)  \
58                  1            17170223                    47   
45                  4            18129115                    49   
46                  5            18129115                    49   
47                  4            18129115                    49   
48                  6            18129115                    49   

    Calculated Exposure (mrem)        City Province/Territory  
58                    0.021277     Camrose            Alberta  
45                    0.081633  Port Dover            Ontario  
46                    0.102041  Port Dover            Ontario  
47                    0.081633  Port Dover            Ontario  
48                    0.122449  Port Dover            Ontario

# 2. Clean the datetime
Remove the characters at the end of date (minutes, timezone).
_NOTE_: Do not run the cell bellow TWICE as it will keep on erasing the end.

In [17]:
# Strip the random endings and leave only the dates alone. No minutes, timezone, etc.

studs_sorted_df['Collection Date']= studs_sorted_df['Collection Date'].str[:-19]
studs_sorted_df.head(5)
# DO NO RERUN THIS CELL

Identifier Collection Date  Number of bubbles  \
58  009d1d48-27ab-4446-b20a-b67855ff8e8b      2019-01-30                  1   
45  02127c7f-69ce-41ee-ae58-402c405b5e49      2019-01-22                  4   
46  02127c7f-69ce-41ee-ae58-402c405b5e49      2019-01-23                  5   
47  02127c7f-69ce-41ee-ae58-402c405b5e49      2019-01-23                  4   
48  02127c7f-69ce-41ee-ae58-402c405b5e49      2019-01-24                  6   

    Bubble Detector ID  Sensitivity (b/mrem)  Calculated Exposure (mrem)  \
58            17170223                    47                    0.021277   
45            18129115                    49                    0.081633   
46            18129115                    49                    0.102041   
47            18129115                    49                    0.081633   
48            18129115                    49                    0.122449   

          City Province/Territory  
58     Camrose            Alberta  
45  Port Dover            Ontario  
46  Port Dover            Ontario  
47  Port Dover            Ontario  
48  Port Dover            Ontario

### Remove unique identifiers
We want to work with data that has been collected at least twice. Otherwise, it is impossible to predict what a 7th day collection would have looked like.

In [18]:
# remove single identifiers
studs_sorted_df.groupby(['Identifier','Collection Date']).filter(lambda x: len(x)>1).shape

(44, 8)

# 3. Reset Index

In [19]:
#check column names
studs_sorted_df.columns

Index(['Identifier', 'Collection Date', 'Number of bubbles',
       'Bubble Detector ID', 'Sensitivity (b/mrem)',
       'Calculated Exposure (mrem)', 'City', 'Province/Territory'],
      dtype='object')

In [20]:
# reset index
studs_sorted_df = studs_sorted_df.reset_index()

In [21]:
studs_sorted_df.drop(columns = 'index').head(5)

Identifier Collection Date  Number of bubbles  \
0  009d1d48-27ab-4446-b20a-b67855ff8e8b      2019-01-30                  1   
1  02127c7f-69ce-41ee-ae58-402c405b5e49      2019-01-22                  4   
2  02127c7f-69ce-41ee-ae58-402c405b5e49      2019-01-23                  5   
3  02127c7f-69ce-41ee-ae58-402c405b5e49      2019-01-23                  4   
4  02127c7f-69ce-41ee-ae58-402c405b5e49      2019-01-24                  6   

   Bubble Detector ID  Sensitivity (b/mrem)  Calculated Exposure (mrem)  \
0            17170223                    47                    0.021277   
1            18129115                    49                    0.081633   
2            18129115                    49                    0.102041   
3            18129115                    49                    0.081633   
4            18129115                    49                    0.122449   

         City Province/Territory  
0     Camrose            Alberta  
1  Port Dover            Ontario  
2  Port Dover            Ontario  
3  Port Dover            Ontario  
4  Port Dover            Ontario

# 4. Save the file externally.

In [22]:
studs_sorted_df.to_csv("students_clean.csv", sep = ',')

Note: The file students_clean got modified into `radiation.csv` by a teammember. The following files can be found in the folder `data` on my Github [here.](https://github.com/rominacarabathampi/CanadianSpaceAppChallenge/tree/master/data)

# 5. Merge file with Radiation data

In [23]:
import pandas as pd
# read radiation file
df = pd.read_csv('radiation.csv')

In [24]:
#check column names
df.columns

Index(['Identifier', 'Collection Date', 'Number of bubbles',
       'Bubble Detector ID', 'Sensitivity (b/mrem)',
       'Calculated Exposure (mrem)', 'City', 'Province/Territory',
       'Unnamed: 8', 'Calculated Exposure Day 7'],
      dtype='object')

### Remove all zeros from the data.

In [25]:
# remove al rows that have values of zeros
df = df[(df.T != 0).all()]

### Change the dataframe to contain only columns of interest.

In [26]:
# select columns Identifier, City, Province/Territory, Calculated Exposure Day 7
df = df.loc[:, ['Identifier','City','Province/Territory','Calculated Exposure Day 7']]

In [27]:
# check columns names after change
df.columns

Index(['Identifier', 'City', 'Province/Territory',
       'Calculated Exposure Day 7'],
      dtype='object')

### Remove NaNs

In [28]:
# Remove all NaNs from the data
df = df.dropna()

### Change Nouvelle-?cosse to Nova Scotia

In [29]:
# Replace Nouvelle-?cosse to Nova Scotia
df.loc[(df['Province/Territory'] == 'Nouvelle-?cosse'),'Province/Territory'] ='Nova Scotia'

In [30]:
# see the dimension of the file

In [31]:
df.shape

(109, 4)

## 6. Read the research file on radiation 
And merge them on cities. This will help comparison of neutron radiation to be compared to all other radiation sources per city.

In [32]:
research = pd.read_csv('City-wise Radiation Information - Feuille 1.csv')
research.head(10)

City  Radon(mrem)  \
0                               Dartmouth       62.900   
1                                 Limoges       62.900   
2                  Whitchurch-Stouffville      188.700   
3  Sainte-Catherine-de-la-Jacques-Cartier      188.700   
4                               Newmarket        1.887   
5                      St-Felix-de-Valois       62.900   
6                             Stouffville      188.700   
7                   St-Jean-sur-Richelieu      314.500   
8                                Oakville      188.700   
9                               Saskatoon      314.500   

   External Terrestrial (mrem)  Cosmic (mrem)  Ingestion (mrem)  
0                         33.3           32.5              30.6  
1                         24.0           35.0              30.6  
2                         17.8           35.0              30.6  
3                         21.9           37.5              30.6  
4                         17.8           35.0              30.6  
5                         27.8           37.5              30.6  
6                         17.8           35.0              30.6  
7                         27.8           37.5              30.6  
8                         17.8           35.0              30.6  
9                         33.3           42.5              30.6

In [33]:
# merge on City
merged = pd.merge(df, research, on=['City'])

In [34]:
# dimensions of merged file
merged.shape

(105, 8)

In [35]:
# column names
merged.columns

Index(['Identifier', 'City', 'Province/Territory', 'Calculated Exposure Day 7',
       'Radon(mrem)', 'External Terrestrial (mrem)', 'Cosmic (mrem)',
       'Ingestion (mrem)'],
      dtype='object')

# 7. Save file externally.

Save the final file externally.

In [36]:
merged.to_csv('merged_files.csv')